lib/tree.rb

```diff
@@ -16,51 +16,93 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def add(key, value)
-    raise NotImplementedError
+  # Time Complexity: O(logn) 
+  # Space Complexity: O(n)
+  def add(key, value, current=@root)
+    # if the root is nil, the tree is empty, add a new node as root
+    return @root = TreeNode.new(key, value) unless @root
+
+    if key < current.key
+      return add(key, value, current.left) if current.left
+      current.left = TreeNode.new(key, value)
+    elsif key > current.key
+      return add(key, value, current.right) if current.right
+      current.right = TreeNode.new(key, value)
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def find(key)
-    raise NotImplementedError
+  # Time Complexity: O(logn) 
+  # Space Complexity: O(n)
+  def find(key, current=@root)
```

Comment:
The space/time complexity is O(log n) if the tree is balanced and O(n) if it's not.